code/modules/space_management/space_transition.dm

```diff
@@ -0,0 +1,412 @@
+/datum/point/proc/get_empty_neighbors(datum/spacewalk_grid/SW)
+	var/list/result
+	var/datum/point/up = SW.get(x, y+1, allow_empty = 1)
+	var/datum/point/down = SW.get(x, y-1, allow_empty = 1)
+	var/datum/point/left = SW.get(x-1, y, allow_empty = 1)
+	var/datum/point/right = SW.get(x+1, y, allow_empty = 1)
+	if(isnull(up))
+		up = new(x, y+1)
+		SW.add_available_node(up)
+
+	if(isnull(down))
+		down = new(x, y-1)
+		SW.add_available_node(down)
+
+	if(isnull(left))
+		left = new(x-1, y)
+		SW.add_available_node(left)
+
+	if(isnull(right))
+		right = new(x+1, y)
+		SW.add_available_node(right)
+
+	// Nodes with a space datum are not empty
+	result = list(up, down, left, right)
+	for(var/datum/point/thing in result)
+		if(!isnull(thing.spl))
+			result -= thing
+	return result
+
+// This looks around itself to see if it has any active nodes within the cardinal directions
+/datum/point/proc/has_no_neighbors(datum/spacewalk_grid/SW)
+	var/result = 1
+	if(spl)
+		return 1
+	if(!isnull(SW.get(x+1,y)))
+		result = 0
+	if(!isnull(SW.get(x-1,y)))
+		result = 0
+	if(!isnull(SW.get(x,y+1)))
+		result = 0
+	if(!isnull(SW.get(x,y-1)))
+		result = 0
+	return result
+
+
+/datum/point/proc/deactivate()
+	if(!spl)
+		throw EXCEPTION("Attempted to deactivate inactive point")
+	for(var/direction in spl.neighbors)
+		var/datum/space_level/S = spl.neighbors[direction]
+		var/oppose = get_opposite_direction(direction)
+		S.neighbors.Remove(oppose)
+	spl.reset_connections()
+	spl = initial(spl)
+
+// This is like the old algorithm, except this one
+// can expand indefinitely and add new points on the fly
+
+// The algorithm is: Start in the center, and add all adjacent points to a list of things to select
+// Then repeatedly do the cycle of choosing a node next to any previously-selected node, then
+// add all points next to your chosen node to the list of things to select
+/datum/spacewalk_grid
+	// This is a list of fully initialized nodes - these have their neighbors
+	// assigned, and are surrounded
+	var/list/filled_nodes = list()
+	// These nodes are not fully initialized - their neighbors field is empty,
+	// and they do not correspond to a space level
+	var/list/available_nodes = list()
+	var/list/all_nodes = list() // filled_nodes | available_nodes
+	var/min_x = 0
+	var/min_y = 0
+	var/max_x = 0
+	var/max_y = 0
+
+/datum/spacewalk_grid/New()
+	var/datum/point/P = new(0,0)
+	add_available_node(P)
+
+/datum/spacewalk_grid/proc/add_available_node(datum/point/P)
+	var/hash = P.hash()
+	if(hash in all_nodes)
+		log_debug("Hash overlap! [hash]")
+	all_nodes[P.hash()] = P
+	available_nodes |= P
+
+// We call this to flag a node as in use - all required variables will be
+// ready after this is called
+/datum/spacewalk_grid/proc/consume_node(datum/point/P)
+	filled_nodes |= P
+	available_nodes -= P
+	// This proc adds any new neighbors to the list of available nodes
+	P.get_empty_neighbors(src)
+	if(min_x > P.x)
+		min_x = P.x
+	else if (max_x < P.x)
```

Comment:
Oh no
I'm going to be devoured by a code formatting nazi

It was nice knowing you